### ISSUE TITLE
guzzle post recv error 56, however the cUrl works. Workaround available

### ISSUE THREAD

**Issue by cfv1000** [guzzle/guzzle#1856]
Hello,

I get cURL error 56: Recv failure when running the following code

```php 
include_once "vendor/autoload.php";

$client = new \GuzzleHttp\Client();

$response = $client->post($url, [
    'form_params' => $post,
]);

print $response->getBody()->getContents();
```

The php cUrl works properly

```php 

$ch = curl_init();
curl_setopt($ch, CURLOPT_URL, $url);
curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch, CURLOPT_POSTFIELDS, $post);
$html = curl_exec($ch);
print $html;
```
So i made some changes to my Guzzle code and it works
```php
include_once "vendor/autoload.php";

$client = new \GuzzleHttp\Client();

$response = $client->post($url, [
    'curl' => [
       CURLOPT_HTTPHEADER => [],
      CURLOPT_POSTFIELDS => $post
    ]
]);

print $response->getBody()->getContents();

````

This seems like a bug to me (on your code). 
The server does not recognize the request as a valid POST

**Comment by sagikazarmark** [comment]
Can you show me an example of your `$post` variable?

>  The server does not recognize the request as a valid POST

Based on what do you make that assumption? Do you have any logs from the server (would help a lot)?

Can you turn on debug mode and paste the output here/compare with what your cURL call sends to the server?

(Sidenote: using `getContents` is not safe as it might not return the full body if something (like a middleware) reads it prior to your call. Casting the body (stream) to string is safer)

**Comment by cfv1000** [comment]
These are simple calls - I could send you the post, however I can't disclose any server side informations ( sorry)

I make that assumption based on the fact that I can make POST requests to the server with workaround, postman and php curl library.

> Based on what do you make that assumption? Do you have any logs from the server (would help a lot)?

On the fact that I can't make POST requests to the server with Guzzle, but I can with php curl library (or by overriding the curl options).

```php
$options = array(
                "curl" => [
                    CURLOPT_HTTPHEADER => [],
                    CURLOPT_POSTFIELDS => [
                        'cod'=> "11054529",
                        'cap'=>null,
                        'B1'=>'VIZUALIZARE'
                    ]
                ]
            );
````

**Comment by sagikazarmark** [comment]
But what makes you think *the server does not recognize the request as a valid POST*?

Do you have any information which can help us track down *why* the server thinks it's a bad request?

Guzzle does a lot of magic which the plain curl extension does not. Could be a wrong combination of headers or anything like that.

Since we don't have any information about the server side I can hardly track down what the problem is. Would be nice if you could do some more digging, maybe you can find the source of the problem.

**Comment by cfv1000** [comment]
As far as I can tell it's a combination of issues:
1st issue: CURLOPT_POSTFIELDS is altered -converted to string, and it removed one of my fields wich i set to ``` null ```.
2nd issue: CURLOPT_HEADERS contains some header that the server just doesn't seem to like (i'm guessing it's the Content-Type, but I haven't tested).

I can give you details in private if you want ( including the server url for testing ), but this can't be publicly posted.

**Comment by sagikazarmark** [comment]
> 1st issue: CURLOPT_POSTFIELDS is altered

Yeah, null fields are removed. Try setting it to an empty string.

> 2nd issue: CURLOPT_HEADERS contains some header that the server just doesn't seem to like

Would be nice to know some details here, but if this is the case, then it's very likely not a Guzzle bug. Setting the Content-Type (and some other headers) is pretty much aligned with HTTP standards, so if the server does not like it, it's rather a server issue.

Anyway, you can try writing a middleware to remove those headers.

**Comment by cfv1000** [comment]
CURL_POST: 1 and CURLOPT_POSTFIELDS: array() are enough. cURL is sending a proper valid HTTP request to a server just by using those 2. Anything else is just overkill. And this overkill is not working in 100% of cases ( as you can see ). 

I already have my workaround and I can work with it. So in my case this is not an issue. It's all up to the maintainer if he wants to make any kind of adjustments to the library.

**Comment by sagikazarmark** [comment]
> cURL is sending a proper valid HTTP request to a server just by using those 2

This does not mean Guzzle sends an invalid request per se. Guzzle tries to comply with accepted HTTP standards, but servers don't always implement them the same way. While cURL is simple to use, Guzzle tries to be useful by putting in some extra logic which should work in most of the cases, but we can't support every server-specific behaviour.

So unless there is something specific which is not valid according to HTTP standards I fear I can't do anything, sorry.

If you can provide me some details why the server thinks the request is invalid (truncated error messages could be enough if they contain the reason) then I might be able to investigate if we can do something or not.

**Comment by talha5389** [comment]
This issue also happened to me.
I integrated Digi-Key api but suddenly it started giving me error like:
`cURL error 56: SSL read: error:00000000:lib(0):func(0):reason(0), errno 0 (see http://curl.haxx.se/libcurl/c/libcurl-errors.html)`

I was using Guzzle 6.3 with curl 7.51. Apache 2.4 x64 with PHP 7.1 x64 in windows. Also, I was using in Lumen Microframework 5.4.

**Another note: although in composer.json my Guzzle version was 6.3, in error it was reporting 6.2.x**

This was my client configuration
`$this->client = new Client([
            'base_uri' => 'https://api.digikey.com/services/',
            'verify' => false
]);`

This was my request

> $formParams = [
            'client_id' => $this->clientID,
            'client_secret' => $this->clientSecret,
            'grant_type' => 'refresh_token',
            'code' => $oAuthAuthorizationCOde
        ];
        $res = $this->client->request('POST', 'https://sso.digikey.com/as/token.oauth2', [
//                'form_params' => $formParams,
            'curl'=>[
                CURLOPT_HTTPHEADER => [],
                CURLOPT_POSTFIELDS => http_build_query($formParams)
            ]
        ]);

As you can see, I was using **form_params** but suddenly it started giving me error. After spending 2-3 hours, I resolved the issue by sending data directly via curl as shown above.

I will gladly send any debug logs but I don't know how to retrieve them. I hope that this issue is identified and resolved

**Comment by kwaiksa** [comment]
I encountered this issue too.
Error like below: `cURL error 56: SSL read: error:00000000:lib(0):func(0):reason(0), errno 10054 (see http://curl.haxx.se/libcurl/c/libcurl-errors.html)`

In debug message, I saw guzzle send `User-Agent: GuzzleHttp/6.2.1 curl/7.50.1 PHP/7.0.10`
So I try to set User-Agent to empty string , "Moliza/5.0", "test"......etc.
Then it's resolved.
In my case, word "PHP" is not allowed for my API server.
